pep-0683.rst

```diff
@@ -0,0 +1,335 @@
+multiple threads that concurrently incref or decref.  Without a shared
+GIL, two running interpreters could not safely share any objects,
+even otherwise immutable ones like ``None``.
+
+This means that, to have a per-interpreter GIL, each interpreter must
+have its own copy of *every* object, including the singletons and
+static types.  We have a viable strategy for that but it will
+require a meaningful amount of extra effort and extra
+complexity.
+
+The alternative is to ensure that all shared objects are truly immutable.
+There would be no races because there would be no modification.  This
+is something that the immortality proposed here would enable for
+otherwise immutable objects.  With immortal objects,
+support for a per-interpreter GIL
+becomes much simpler.
+
+Avoiding Copy-on-Write
+----------------------
+
+For some applications it makes sense to get the application into
+a desired initial state and then fork the process for each worker.
+This can result in a large performance improvement, especially
+memory usage.  Several enterprise Python users (e.g. Instagram,
+YouTube) have taken advantage of this.  However, the above
+refcount semantics drastically reduce the benefits and
+has led to some sub-optimal workarounds.
+
+Also note that "fork" isn't the only operating system mechanism
+that uses copy-on-write semantics.
+
+
+Rationale
+=========
+
+The proposed solution is obvious enough that two people came to the
+same conclusion (and implementation, more or less) independently.
+Other designs were also considered.  Several possibilities
+have also been discussed on python-dev in past years.
+
+Alternatives include:
+
+* use a high bit to mark "immortal" but do not change ``Py_INCREF()``
+* add an explicit flag to objects
+* implement via the type (``tp_dealloc()`` is a no-op)
+* track via the object's type object
+* track with a separate table
+
+Each of the above makes objects immortal, but none of them address
+the performance penalties from refcount modification described above.
+
+In the case of per-interpreter GIL, the only realistic alternative
+is to move all global objects into ``PyInterpreterState`` and add
+one or more lookup functions to access them.  Then we'd have to
+add some hacks to the C-API to preserve compatibility for the
+may objects exposed there.  The story is much, much simpler
+with immortal objects
+
+
+Impact
+======
+
+Benefits
+--------
+
+Most notably, the cases described in the two examples above stand
+to benefit greatly from immortal objects.  Projects using pre-fork
+can drop their workarounds.  For the per-interpreter GIL project,
+immortal objects greatly simplifies the solution for existing static
+types, as well as objects exposed by the public C-API.
+
+Performance
+-----------
+
+A naive implementation shows `a 4% slowdown`_.
+Several promising mitigation strategies will be pursued in the effort
+to bring it closer to performance-neutral.
```

Comment:
Probably not needed to include here but these are four different strategies that I'll try:
* Immortalized Interned Strings (including Py_IDENTIFIERs)
* Immortalizing heap after startup
* GC improvements for large applications
* Immortalizing Module-level globals at import time